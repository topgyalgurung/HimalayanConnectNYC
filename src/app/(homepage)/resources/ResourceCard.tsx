/**
 * ResourceCard Component
 *
 * This component is the resource card for the homepage of the Himalayan Connect NYC application.
 * It displays a single resource in a card format.
 *
 */

"use client";

import { format } from "date-fns";

import { type Resource } from "@/app/types/resource";
import { formatOpenDays } from "@/app/lib/helpers/formatOpenDays";

interface ResourceCardProps {
  resources: Resource[];
  onViewDetails?: (resource: Resource) => void;
}
// ResourceCard component: Displays a list of resources
export default function ResourceCard({
  resources,
  onViewDetails,
}: ResourceCardProps) {
  return (
    <div className="flex flex-col space-y-4 pb-20">
      {resources
        .filter((resource) => resource.status === "APPROVED")
        .map((resource) => (
          <div
            key={resource.id}
            className="flex justify-between p-4 border rounded-md shadow-md bg-white transition-all duration-300 hover:shadow-xl hover:-translate-y-1 hover:scale-[1.02]"
          >
            <div className="flex-1">
              <h4 className="text-lg font-bold">{resource.name}</h4>
              <p className="text-gray-500">
                {resource.ResourceCategory?.name || "no category"}
              </p>
              <p>{resource.description}</p>
              <p>
                <strong>Borough: </strong> {resource.city}
              </p>
              <p>
                <strong>Address: </strong>
                {resource.address ? resource.address : "No address available"}
              </p>
              <p>
                <strong>Open Days :</strong>{" "}
                {resource.openDays ? formatOpenDays(resource.openDays) : "N/A"}
              </p>
              <p className="flex items-center gap-1">
                <strong>Hours:</strong>
                <span className="text-gray-700">
                  {resource.openTime
                    ? format(resource.openTime, "hh:mm a")
                    : "N/A"}{" "}
                  -{" "}
                  {resource.closeTime
                    ? format(resource.closeTime, "hh:mm a")
                    : "N/A"}
                </span>
              </p>
            </div>

            <div className="flex items-end ml-4">
              <button
                className="text-white py-2 px-3 bg-blue-600 rounded hover:bg-blue-700"
                onClick={() => onViewDetails?.(resource)}
              >
                View Details
              </button>
            </div>
          </div>
        ))}
    </div>
  );
}
