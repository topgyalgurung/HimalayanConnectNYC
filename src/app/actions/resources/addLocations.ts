import prisma from "@/app/lib/prisma";

// Coordinates for NYC boroughs
const BOROUGH_COORDINATES = {
  "Manhattan": { lat: 40.7831, lng: -73.9712 },
  "Brooklyn": { lat: 40.6782, lng: -73.9442 },
  "Queens": { lat: 40.7282, lng: -73.7949 },
  "Bronx": { lat: 40.8448, lng: -73.8648 },
  "Staten Island": { lat: 40.5795, lng: -74.1502 }
};

export async function addLocations() {
  try {
    // Get all resources without locations
    const resources = await prisma.resource.findMany({
      where: {
        Location: {
          none: {}
        }
      },
      select: {
        id: true,
        name: true,
        city: true,
        address: true
      }
    });

    console.log(`Found ${resources.length} resources without locations`);

    // Add locations for each resource
    for (const resource of resources) {
      // Try to determine coordinates based on borough
      const coordinates = BOROUGH_COORDINATES[resource.city as keyof typeof BOROUGH_COORDINATES] || 
                         BOROUGH_COORDINATES["Manhattan"]; // Default to Manhattan if borough not found

      // Create location
      await prisma.location.create({
        data: {
          resourceId: resource.id,
          latitude: coordinates.lat,
          longitude: coordinates.lng
        }
      });

      console.log(`Added location for resource ${resource.id} (${resource.name})`);
    }

    return { success: true, message: `Added locations for ${resources.length} resources` };
  } catch (error) {
    console.error("Error adding locations:", error);
    return { success: false, error: "Failed to add locations" };
  }
} 