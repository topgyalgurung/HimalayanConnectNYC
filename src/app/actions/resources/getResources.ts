/**
 * Server Action for Fetching Resources
 * 
 * This server action fetches resources from the database with proper filtering
 * and includes related data like categories, locations, and user likes.
 */

import { prisma } from "@/app/lib/prisma";
import { getSession } from "@/app/lib/session";
import type { Resource } from "@/app/lib/types";
import { ResourceStatus } from "@prisma/client";

interface GetResourcesOptions {
  categories?: string[];
  boroughs?: string[];
}

export async function getResources({ categories = [], boroughs = [] }: GetResourcesOptions = {}): Promise<Resource[]> {
  try {
    const session = await getSession();
    const userId = session?.userId ? Number(session.userId) : null;

    // Build where clause based on filters
    const where = {
      ...(categories && categories.length > 0 && {
        categoryId: {
          in: categories.map(Number)
        }
      }),
      ...(boroughs && boroughs.length > 0 && {
        city: {
          in: boroughs
        }
      }),
      status: ResourceStatus.APPROVED // Only show approved resources
    };

    const resources = await prisma.resource.findMany({
      where,
      include: {
        ResourceCategory: {
          select: { id: true, name: true },
        },
        Location: {
          select: {
            id: true,
            resourceId: true,
            latitude: true,
            longitude: true,
          },
        },
        ResourceLike: userId
          ? {
              where: { userId },
              select: { id: true },
            }
          : false,
      },
    });

    // Serialize all Decimal fields to numbers and convert IDs to strings
    return resources.map(resource => ({
      ...resource,
      id: String(resource.id),
      rating: Number(resource.rating),
      Location: resource.Location.map(loc => ({
        ...loc,
        latitude: loc.latitude ? Number(loc.latitude) : null,
        longitude: loc.longitude ? Number(loc.longitude) : null,
      })),
      createdAt: resource.createdAt,
      openTime: resource.openTime || null,
      closeTime: resource.closeTime || null,
      ResourceCategory: resource.ResourceCategory ? {
        ...resource.ResourceCategory,
        id: String(resource.ResourceCategory.id),
      } : null,
    }));
  } catch (error) {
    console.error("Error fetching resources:", error);
    throw new Error("Failed to fetch resources");
  }
} 