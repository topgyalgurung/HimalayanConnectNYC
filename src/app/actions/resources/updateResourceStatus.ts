'use server';

import { revalidatePath } from 'next/cache';
import { prisma } from '@/app/lib/prisma';
import { ResourceStatus, Resource } from '@prisma/client';

type ResourceUpdateData = Partial<Pick<Resource, 'name' | 'address' | 'phone' | 'url' | 'openDays' | 'openTime' | 'closeTime'>>;

// this is the action for the admin to update the status of the resource
// todo: when edit status is approved, update the resource with the new values
export async function updateResourceStatus(
  resourceId: string,
  newStatus: ResourceStatus,
  resourceType: 'new' | 'edit'
) {
  try {
    if (!["PENDING", "APPROVED", "REJECTED"].includes(newStatus)) {
      return { success: false, error: "Invalid status" };
    }

    const resourceIdNum = parseInt(resourceId);
    if (isNaN(resourceIdNum)) {
      return { success: false, error: "Invalid resource ID" };
    }

    if (resourceType === "edit") {
      // First fetch the edit suggestion
      const resourceEditSuggestion = await prisma.resourceEditSuggestion.findUnique({
        where: { id: resourceIdNum }
      });

      if (!resourceEditSuggestion) {
        return { success: false, error: "Edit suggestion not found" };
      }

      await prisma.resourceEditSuggestion.update({
        where: { id: resourceIdNum },
        data: { status: newStatus },
      });

      // If the suggestion is approved, update the resource
      if (newStatus === "APPROVED") {
        // Create an update object with only the fields that have values
        const updateData: ResourceUpdateData = {};
        
        if (resourceEditSuggestion.name) updateData.name = resourceEditSuggestion.name;
        if (resourceEditSuggestion.address) updateData.address = resourceEditSuggestion.address;
        if (resourceEditSuggestion.phone) updateData.phone = resourceEditSuggestion.phone;
        if (resourceEditSuggestion.url) updateData.url = resourceEditSuggestion.url;
        if (resourceEditSuggestion.openDays) updateData.openDays = resourceEditSuggestion.openDays;
        if (resourceEditSuggestion.openTime) updateData.openTime = resourceEditSuggestion.openTime;
        if (resourceEditSuggestion.closeTime) updateData.closeTime = resourceEditSuggestion.closeTime;

        console.log('Updating resource with data:', updateData);

        // Update the resource with the changed fields
        await prisma.resource.update({
          where: { id: resourceEditSuggestion.resourceId },
          data: updateData
        });
      }
    } else {
      await prisma.resource.update({
        where: { id: resourceIdNum },
        data: { status: newStatus },
      });
    }

    // Revalidate the paths that might be affected
    revalidatePath('/profile/admin');
    revalidatePath('/api/resources');
    revalidatePath('/api/resources/edit');

    return { success: true };
  } catch (error) {
    console.error("Error updating resource status:", error);
    return { 
      success: false, 
      error: "Failed to update resource status" 
    };
  }
}
