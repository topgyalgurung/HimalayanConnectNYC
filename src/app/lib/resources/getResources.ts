/**
 * Server Action for Fetching Resources
 * 
 * This server action fetches resources from the database with proper filtering
 * and includes related data like categories, locations, and user likes.
 */

// todo: move both into one resource.ts 

import { prisma } from "@/app/lib/prisma";
import { getSession } from "@/app/lib/session";
import type { Resource } from "@/app/lib/definitions";
import { ResourceStatus } from "@prisma/client";

type GetResourcesOptions = {
  categories?: number[];
  boroughs?: string[];
  query?: string;
  currentPage?: number;
};

// getResources

export async function getResources({ categories = [], boroughs = [] }: GetResourcesOptions = {}): Promise<Resource[]> {
  try {
    const session = await getSession();
    const userId = session?.userId ? Number(session.userId) : null;

    // Build where clause based on filters
    const where = {
      ...(categories && categories.length > 0 && {
        categoryId: {
          in: categories.map(Number)
        }
      }),
      ...(boroughs && boroughs.length > 0 && {
        city: {
          in: boroughs
        }
      }),
      status: ResourceStatus.APPROVED // Only show approved resources
    };

    const resources = await prisma.resource.findMany({
      where,
      include: {
        ResourceCategory: {
          select: { id: true, name: true },
        },
        Location: {
          select: {
            id: true,
            resourceId: true,
            latitude: true,
            longitude: true,
          },
        },
        ResourceLike: userId
          ? {
              where: { userId },
              select: { id: true },
            }
          : false,
      },
    });

    // Serialize all Decimal fields to numbers and convert IDs to strings
    return resources.map(resource => ({
      ...resource,
      id: String(resource.id),
      rating: Number(resource.rating),
      Location: resource.Location.map(loc => ({
        ...loc,
        latitude: loc.latitude ? Number(loc.latitude) : null,
        longitude: loc.longitude ? Number(loc.longitude) : null,
      })),
      createdAt: resource.createdAt,
      openTime: resource.openTime || null,
      closeTime: resource.closeTime || null,
      ResourceCategory: resource.ResourceCategory ? {
        ...resource.ResourceCategory,
        id: String(resource.ResourceCategory.id),
      } : null,
    }));
  } catch (error) {
    console.error("Error fetching resources:", error);
    throw new Error("Failed to fetch resources");
  }
} 



const ITEMS_PER_PAGE = 6;


export async function fetchFilteredResources({
  categories = [],
  boroughs = [],
  query = "",
  currentPage = 1,
}: GetResourcesOptions = {}): Promise<Resource[]> {
  try {
    const session = await getSession();
    const userId = session?.userId ? Number(session.userId) : null;

    // Calculate pagination
    const skip = (currentPage - 1) * ITEMS_PER_PAGE;

    const where = {
      ...(categories && categories.length > 0 && {
        categoryId: {
          in: categories.map(Number)
        }
      }),
      ...(boroughs && boroughs.length > 0 && {
        city: {
          in: boroughs
        }
      }),
      status: ResourceStatus.APPROVED // Only show approved resources
    };

    const resources = await prisma.resource.findMany({
      where,
      skip,
      take: ITEMS_PER_PAGE,
      include: {
        ResourceCategory: {
          select: { id: true, name: true },
        },
        Location: {
          select: {
            id: true,
            resourceId: true,
            latitude: true,
            longitude: true,
          },
        },
        ResourceLike: userId
          ? {
              where: { userId },
              select: { id: true },
            }
          : false,  
      },
    });

    // Transform the data to match the Resource type
    return resources.map(resource => {
      const locations: Location[] = resource.Location.map(loc => ({
        id: Number(loc.id),
        latitude: Number(loc.latitude),
        longitude: Number(loc.longitude),
      }));

      return {
        id: String(resource.id),
        name: resource.name,
        description: resource.description || null,
        address: resource.address || null,
        city: resource.city || null,
        phone: resource.phone || null,
        email: resource.email || null,
        url: resource.url || null,
        openDays: resource.openDays || null,
        openTime: resource.openTime || null,
        closeTime: resource.closeTime || null,
        status: resource.status,
        createdAt: resource.createdAt,
        categoryId: resource.categoryId || null,
        rating: resource.rating ? Number(resource.rating) : null,
        imageUrl: resource.imageUrl || null,
        facebookLink: resource.facebookLink || null,
        ResourceCategory: resource.ResourceCategory ? {
          name: resource.ResourceCategory.name,
        } : null,
        Location: locations,
      };
    });
  } catch (error) {
    console.error("Failed to fetch resources:", error);
    throw new Error("Unable to fetch resources.");
  }
}

export async function fetchResourcesPages({
  categories = [],
  boroughs = [],
  query = "",
}: GetResourcesOptions = {}): Promise<number> {
  try {
    const where: Prisma.ResourceWhereInput = {
      ...(categories.length > 0 && {
        categoryId: { in: categories.map(Number) },
      }),
      ...(boroughs.length > 0 && {
        city: { in: boroughs },
      }),
      status: ResourceStatus.APPROVED,
      ...(query && {
        OR: [
          { name: { contains: query, mode: 'insensitive' } },
          { city: { contains: query, mode: 'insensitive' } },
          { description: { contains: query, mode: 'insensitive' } },
        ],
      }),
    };

    const count = await prisma.resource.count({ where });
    const totalPages = Math.ceil(count / ITEMS_PER_PAGE);
    return totalPages;
  } catch (error) {
    console.error("Failed to count resources:", error);
    throw new Error("Unable to count matching resources.");
  }
}