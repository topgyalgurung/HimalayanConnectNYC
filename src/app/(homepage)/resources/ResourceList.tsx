/**
 * resource list for the homepage of the Himalayan Connect NYC application.
 * It displays the list of resources in a sidebar.
 * 
 */

"use client";
import { useSearchParams } from "next/navigation";
// import { useEffect } from "react";
import ResourceCard from "./ResourceCard";
import type { Resource } from "@/app/lib/types";

interface ResourceListProps {
  filteredResources: Resource[];
  onViewDetailsAction: (resource: Resource) => void;
}

const ResourceList = ({
  filteredResources,
  onViewDetailsAction,
}: ResourceListProps) => {
  const searchParams = useSearchParams();
  const searchQuery = searchParams.get("query")?.toLowerCase() || "";

  // Filter resources based on search query only
  const searchFilteredResources = filteredResources.filter((resource) => {
    const searchQueryLower = searchQuery.trim().toLowerCase();
    const nameMatch = resource.name?.toLowerCase().includes(searchQueryLower);
    const addressMatch = resource.address
      ?.toLowerCase()
      .includes(searchQueryLower);
    const locationMatch = resource.city
      ?.toLowerCase()
      .includes(searchQueryLower);

      return nameMatch || addressMatch || locationMatch;

    // // Check if the resource belongs to a selected category (or allow all if none selected)
    // const categoryMatch =
    //   selectedCategories.length === 0 || // Allow all if no category is selected
    //   selectedCategories.includes(
    //     (resource.ResourceCategory?.name ?? "").toLowerCase()
    //   );

    // // Check if the resource is in a selected borough (or allow all if none selected)
    // const boroughMatch =
    //   selectedBoroughs.length === 0 ||
    //   selectedBoroughs.includes(resource.city ?? "");

    // // A resource is included if it matches name/address and selected category and borough
    // return (
    //   (nameMatch || addressMatch || locationMatch) &&
    //   categoryMatch &&
    //   boroughMatch
    // );
  });

  return (
    <div className="flex flex-col h-full">
      <ResourceCard
        resources={searchFilteredResources}
        onViewDetailsAction={onViewDetailsAction}
      />
    </div>
  );
};

export default ResourceList;
