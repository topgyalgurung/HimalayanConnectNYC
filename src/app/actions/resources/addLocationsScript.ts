import { addLocations } from "./addLocations";

async function main() {
  console.log("Starting to add locations...");
  const result = await addLocations();
  console.log("Result:", result);
  process.exit(0);
}

main().catch((error) => {
  console.error("Error running script:", error);
  process.exit(1);
}); 