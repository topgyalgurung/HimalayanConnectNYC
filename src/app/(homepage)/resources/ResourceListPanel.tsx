// src/app/(homepage)/resources/ResourceListPanel.tsx
/**
 * ResourceListPanel Component
 *
 * This component is the resource list panel for the homepage of the Himalayan Connect NYC application.
 * It displays the list of resources in a sidebar.
 * 
 */

"use client";

import ResourceList from "./ResourceList";
import type { Resource } from "@/app/lib/types";

interface ResourceListPanelProps {
  filteredResources: Resource[];
  onViewDetails: (resource: Resource) => void;
}

export default function ResourceListPanel({
  filteredResources,
  onViewDetails,
}: ResourceListPanelProps) {
  return (
    <aside className="w-[30%] pl-4 flex-1 flex flex-col min-h-0 mb-4">
      <h3 className="text-lg px-4 text-center font-bold text-white bg-blue-500 mb-2 sticky top-0 z-10 p-2 shadow">
        RESOURCES
      </h3>
      <main className="flex-1 bg-gray-50 p-4 overflow-y-auto mb-4">
        <ResourceList
          filteredResources={filteredResources}
          onViewDetails={onViewDetails}
        />
      </main>
    </aside>
  );
}
