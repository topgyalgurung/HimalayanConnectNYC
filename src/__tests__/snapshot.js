import { render } from "@testing-library/react";
import Home from "../app/page";
import { getResources } from "../app/actions/resources/getResources";

// Mock the getResources function
jest.mock("../app/actions/resources/getResources", () => ({
  getResources: jest.fn(),
}));

// Mock the HomeClient component
jest.mock("../app/(homepage)/HomeClient", () => {
  return function MockHomeClient({ initialResources }) {
    return (
      <div>
        <h1>Home</h1>
        <div>Resources: {initialResources.length}</div>
      </div>
    );
  };
});

it("renders correctly", async () => {
  // Mock the resources data
  const mockResources = [
    { id: 1, title: "Resource 1" },
    { id: 2, title: "Resource 2" },
  ];
  getResources.mockResolvedValue(mockResources);

  const { container } = render(await Home());
  expect(container).toMatchSnapshot();
});
