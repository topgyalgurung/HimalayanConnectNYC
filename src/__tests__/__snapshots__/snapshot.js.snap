// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div>
    <h1>
      Home
    </h1>
    <div>
      Resources: 
      2
    </div>
  </div>
</div>
`;
